File: android/app/src/main/java/com/kryptos/android/core/LetterStego.kt
package com.kryptos.android.core


object LetterStego {
    private val MAGIC = byteArrayOf(0xC5.toByte(), 0x3A)
    private const val MIN_FRAME_BYTES = 5
    private const val PREFILTER_CHARS = 24
    private const val SCAN_WINDOW = 256

    private class Alphabet(source: String, val blockBytes: Int) {
        val letters: CharArray = source.toCharArray()
        val base: Int = letters.size
        val index: Map<Char, Int> = letters.withIndex().associate { (i, c) -> c to i }
        val charsForBytes = IntArray(blockBytes + 1)
        val bytesForChars = HashMap<Int, Int>()
        val limits = LongArray(blockBytes + 1)
        val blockChars: Int
        val minChars: Int

        private val first: Char = letters.min()
        private val member = BooleanArray(letters.max() - first + 1).also { flags ->
            for (c in letters) flags[c - first] = true
        }

        fun holds(c: Char): Boolean {
            val i = c - first
            return i >= 0 && i < member.size && member[i]
        }

        init {
            check(base >= 2) { "letter alphabet is too small" }
            check(index.size == base) { "letter alphabet has duplicates" }
            for (r in 1..blockBytes) {
                var limit = 1L
                repeat(r) { limit *= 256L }
                limits[r] = limit
                var chars = 0
                var capacity = 1L
                while (capacity < limit) {
                    capacity *= base.toLong()
                    chars++
                }
                charsForBytes[r] = chars
                if (r < blockBytes) bytesForChars[chars] = r
            }
            blockChars = charsForBytes[blockBytes]
            check(!bytesForChars.containsKey(blockChars)) { "tail length collides with a full block" }
            val rest = MIN_FRAME_BYTES % blockBytes
            minChars = (MIN_FRAME_BYTES / blockBytes) * blockChars +
                if (rest == 0) 0 else charsForBytes[rest]
        }
    }

    private val russian by lazy { Alphabet("абвгдежзийклмнопрстуфхцчшщъыьэюя", 5) }
    private val english by lazy { Alphabet("abcdefghijklmnopqrstuvwxyz", 7) }
    private val chinese by lazy { Alphabet("一不与丢个中为久么之乔也买了事于些什仍从令以们份但住使信俩借假停像先克全关其再写冷刚别到前办加却又另只叫号吃名后向吧听吹呀呆呢和哇哈哎哦哪哭唱啊啥啦喂喝嗨嘛嘿回因在坏块多够太她好如妳字它完家对将少就已带年弧当往很得德心忘忙快怎怕怪总恨恩您想懂成我或戴所才扔把抓抱拍拜拿挂指按挺换掉提搞搬救敢斯新於无时是晚更曾替最月有未来查梦欠次正此每比水求没滚点热爸特猜猫玩球用由画留疼病盯盾看真着睡瞧碰祝神离秒穿站笑第等算米糟约给美老而耶能脚脸腿臭船花若著街被要见让讲读谁谈谢赢跑跟躲躺车达还这远连送选都酒里金错长问间陪饿高鱼", 1) }

    private val tables by lazy { listOf(russian, english, chinese) }

    private fun table(language: StegoLanguage): Alphabet =
        when (language) {
            StegoLanguage.RUSSIAN -> russian
            StegoLanguage.CHINESE -> chinese
            StegoLanguage.ENGLISH, StegoLanguage.GERMAN -> english
        }

    internal fun alphabetCharacters(language: StegoLanguage): CharArray = table(language).letters

    fun encode(data: ByteArray, language: StegoLanguage = StegoLanguage.forSystem()): String? =
        StegoSafety.firstCleanCover(data.size) { seed -> encode(data, language, seed) }

    fun encode(data: ByteArray, language: StegoLanguage, seed: Int): String {
        require(data.isNotEmpty()) { "letter stego needs a payload" }
        val inner = ByteArray(MAGIC.size + data.size + 1)
        MAGIC.copyInto(inner, 0)
        data.copyInto(inner, MAGIC.size)
        inner[inner.size - 1] = crc8(inner, 0, inner.size - 1)

        val framed = ByteArray(1 + inner.size)
        framed[0] = (seed and 0xFF).toByte()
        var x = seed and 0xFF
        for (i in inner.indices) {
            x = (x * 197 + 91) and 0xFF
            framed[i + 1] = (inner[i].toInt() xor x).toByte()
        }
        return pack(framed, table(language))
    }

    fun decode(text: String): ByteArray? {
        for (token in StegoTokenizer.runs(text)) {
            for (a in tables) {
                if (token.length < a.minChars) continue
                if (!a.index.containsKey(token[0]) || !structurallyValid(token.length, a)) continue
                val framed = unpack(token, a) ?: continue
                val payload = unframe(framed)
                if (payload != null) return payload
            }
        }
        return null
    }

    private fun structurallyValid(count: Int, a: Alphabet): Boolean {
        val tail = count % a.blockChars
        return tail == 0 || a.bytesForChars.containsKey(tail)
    }

    fun looksLikeStego(text: String): Boolean = decode(text) != null

    fun mightBeStego(text: String): Boolean {
        val limit = minOf(text.length, SCAN_WINDOW)
        val alphabets = tables
        val runs = IntArray(alphabets.size)
        for (i in 0 until limit) {
            val c = text[i].lowercaseChar()
            for (k in alphabets.indices) {
                if (alphabets[k].holds(c)) {
                    if (++runs[k] >= PREFILTER_CHARS) return true
                } else {
                    runs[k] = 0
                }
            }
        }
        return false
    }

    private fun pack(bytes: ByteArray, a: Alphabet): String {
        val out = StringBuilder(bytes.size * a.blockChars / a.blockBytes + a.blockChars)
        var i = 0
        while (i < bytes.size) {
            val r = minOf(a.blockBytes, bytes.size - i)
            val c = a.charsForBytes[r]
            var value = 0L
            for (k in 0 until r) value = (value shl 8) or (bytes[i + k].toLong() and 0xFF)
            val digits = IntArray(c)
            var k = c - 1
            while (k >= 0) {
                digits[k] = (value % a.base).toInt()
                value /= a.base
                k--
            }
            for (d in digits) out.append(a.letters[d])
            i += r
        }
        return out.toString()
    }

    private fun unpack(token: String, a: Alphabet): ByteArray? {
        val out = ByteArray(token.length * a.blockBytes / a.blockChars + a.blockBytes)
        var written = 0
        var i = 0
        while (i < token.length) {
            val remaining = token.length - i
            val c: Int
            val r: Int
            if (remaining >= a.blockChars) {
                c = a.blockChars
                r = a.blockBytes
            } else {
                r = a.bytesForChars[remaining] ?: return null
                c = remaining
            }
            var value = 0L
            for (k in 0 until c) {
                val digit = a.index[token[i + k]] ?: return null
                value = value * a.base + digit
            }
            if (value >= a.limits[r]) return null
            var k = r - 1
            while (k >= 0) {
                out[written++] = ((value shr (8 * k)) and 0xFF).toByte()
                k--
            }
            i += c
        }
        return if (written == out.size) out else out.copyOf(written)
    }

    private fun unframe(framed: ByteArray): ByteArray? {
        if (framed.size < MAGIC.size + 3) return null
        var x = framed[0].toInt() and 0xFF
        val inner = ByteArray(framed.size - 1)
        for (i in 1 until framed.size) {
            x = (x * 197 + 91) and 0xFF
            inner[i - 1] = (framed[i].toInt() xor x).toByte()
        }
        for (i in MAGIC.indices) if (inner[i] != MAGIC[i]) return null
        if (inner[inner.size - 1] != crc8(inner, 0, inner.size - 1)) return null
        return inner.copyOfRange(MAGIC.size, inner.size - 1)
    }

    private fun crc8(data: ByteArray, from: Int, to: Int): Byte {
        var crc = 0
        for (i in from until to) {
            crc = crc xor (data[i].toInt() and 0xFF)
            repeat(8) {
                crc = if (crc and 0x80 != 0) ((crc shl 1) xor 0x07) and 0xFF else (crc shl 1) and 0xFF
            }
        }
        return crc.toByte()
    }
}

File: CipherCore/Sources/CipherCore/Steganography/LetterStego.swift
import Foundation

public enum LetterStego {
    private static let magic: [UInt8] = [0xC5, 0x3A]
    private static let minFrameBytes = 5

    private struct Alphabet: Sendable {
        let letters: [Character]
        let index: [Character: Int]
        let base: Int
        let blockBytes: Int
        let blockChars: Int
        let charsForBytes: [Int]
        let bytesForChars: [Int: Int]
        let limits: [Int]
        let minChars: Int
    }

    private static func alphabet(_ source: String, blockBytes: Int) -> Alphabet {
        let letters = Array(source)
        let base = letters.count
        precondition(base >= 2, "letter alphabet is too small")
        precondition(Set(letters).count == base, "letter alphabet has duplicates")
        var index = [Character: Int](minimumCapacity: base)
        for (i, c) in letters.enumerated() { index[c] = i }

        var charsForBytes = [Int](repeating: 0, count: blockBytes + 1)
        var bytesForChars = [Int: Int](minimumCapacity: blockBytes)
        var limits = [Int](repeating: 0, count: blockBytes + 1)
        for r in 1 ... blockBytes {
            var limit = 1
            for _ in 0 ..< r { limit *= 256 }
            limits[r] = limit
            var chars = 0
            var capacity = 1
            while capacity < limit {
                capacity *= base
                chars += 1
            }
            charsForBytes[r] = chars
            if r < blockBytes { bytesForChars[chars] = r }
        }
        let blockChars = charsForBytes[blockBytes]
        precondition(bytesForChars[blockChars] == nil, "tail length collides with a full block")
        let rest = minFrameBytes % blockBytes
        let minChars = (minFrameBytes / blockBytes) * blockChars + (rest == 0 ? 0 : charsForBytes[rest])
        return Alphabet(letters: letters, index: index, base: base,
                        blockBytes: blockBytes, blockChars: blockChars,
                        charsForBytes: charsForBytes, bytesForChars: bytesForChars, limits: limits,
                        minChars: minChars)
    }

    private static let russian = alphabet("абвгдежзийклмнопрстуфхцчшщъыьэюя", blockBytes: 5)
    private static let english = alphabet("abcdefghijklmnopqrstuvwxyz", blockBytes: 7)
    private static let chinese = alphabet("一不与丢个中为久么之乔也买了事于些什仍从令以们份但住使信俩借假停像先克全关其再写冷刚别到前办加却又另只叫号吃名后向吧听吹呀呆呢和哇哈哎哦哪哭唱啊啥啦喂喝嗨嘛嘿回因在坏块多够太她好如妳字它完家对将少就已带年弧当往很得德心忘忙快怎怕怪总恨恩您想懂成我或戴所才扔把抓抱拍拜拿挂指按挺换掉提搞搬救敢斯新於无时是晚更曾替最月有未来查梦欠次正此每比水求没滚点热爸特猜猫玩球用由画留疼病盯盾看真着睡瞧碰祝神离秒穿站笑第等算米糟约给美老而耶能脚脸腿臭船花若著街被要见让讲读谁谈谢赢跑跟躲躺车达还这远连送选都酒里金错长问间陪饿高鱼", blockBytes: 1)

    private static let persian = alphabet("ابپتثجچحخدذرزژسشصضطظعغفقکگلمنوهی", blockBytes: 5)

    private static let tables = [russian, english, chinese, persian]

    private static func table(for language: StegoLanguage) -> Alphabet {
        switch language {
        case .russian: return russian
        case .chinese: return chinese
        case .persian: return persian
        case .english, .german: return english
        }
    }

    static func alphabetCharacters(for language: StegoLanguage) -> [Character] {
        table(for: language).letters
    }

    public static func encode(_ data: Data, language: StegoLanguage = .forSystem()) -> String? {
        StegoSafety.firstCleanCover(payload: data.count) { seed in
            encode(data, language: language, seed: seed)
        }
    }

    static func encode(_ data: Data, language: StegoLanguage, seed: Int) -> String {
        precondition(!data.isEmpty, "letter stego needs a payload")
        var inner = [UInt8]()
        inner.reserveCapacity(magic.count + 1 + data.count)
        inner.append(contentsOf: magic)
        inner.append(contentsOf: data)
        inner.append(crc8(inner, 0, inner.count))

        var framed = [UInt8]()
        framed.reserveCapacity(1 + inner.count)
        framed.append(UInt8(seed & 0xFF))
        var x = seed & 0xFF
        for byte in inner {
            x = (x * 197 + 91) & 0xFF
            framed.append(byte ^ UInt8(x))
        }
        return pack(framed, table(for: language))
    }

    public static func decode(_ text: String) -> Data? {
        for token in tokenize(text) {
            let chars = Array(token)
            for a in tables {
                guard chars.count >= a.minChars, a.index[chars[0]] != nil,
                      structurallyValid(chars.count, a) else { continue }
                guard let framed = unpack(chars, a) else { continue }
                if let payload = unframe(framed) { return payload }
            }
        }
        return nil
    }

    private static func structurallyValid(_ count: Int, _ a: Alphabet) -> Bool {
        let tail = count % a.blockChars
        return tail == 0 || a.bytesForChars[tail] != nil
    }

    public static func looksLikeStego(_ text: String) -> Bool { decode(text) != nil }

    private static func pack(_ bytes: [UInt8], _ a: Alphabet) -> String {
        var out = ""
        out.reserveCapacity(bytes.count * a.blockChars / a.blockBytes + a.blockChars)
        var i = 0
        while i < bytes.count {
            let r = min(a.blockBytes, bytes.count - i)
            let c = a.charsForBytes[r]
            var value = 0
            for k in 0 ..< r { value = (value << 8) | Int(bytes[i + k]) }
            var digits = [Int](repeating: 0, count: c)
            var k = c - 1
            while k >= 0 {
                digits[k] = value % a.base
                value /= a.base
                k -= 1
            }
            for d in digits { out.append(a.letters[d]) }
            i += r
        }
        return out
    }

    private static func unpack(_ chars: [Character], _ a: Alphabet) -> [UInt8]? {
        var out = [UInt8]()
        out.reserveCapacity(chars.count * a.blockBytes / a.blockChars + a.blockBytes)
        var i = 0
        while i < chars.count {
            let remaining = chars.count - i
            let c: Int
            let r: Int
            if remaining >= a.blockChars {
                c = a.blockChars
                r = a.blockBytes
            } else {
                guard let bytes = a.bytesForChars[remaining] else { return nil }
                c = remaining
                r = bytes
            }
            var value = 0
            for k in 0 ..< c {
                guard let digit = a.index[chars[i + k]] else { return nil }
                value = value * a.base + digit
            }
            guard value < a.limits[r] else { return nil }
            var k = r - 1
            while k >= 0 {
                out.append(UInt8((value >> (8 * k)) & 0xFF))
                k -= 1
            }
            i += c
        }
        return out
    }

    private static func unframe(_ framed: [UInt8]) -> Data? {
        guard framed.count >= magic.count + 3 else { return nil }
        var x = Int(framed[0])
        var inner = [UInt8]()
        inner.reserveCapacity(framed.count - 1)
        for i in 1 ..< framed.count {
            x = (x * 197 + 91) & 0xFF
            inner.append(framed[i] ^ UInt8(x))
        }
        for (i, byte) in magic.enumerated() where inner[i] != byte { return nil }
        guard inner[inner.count - 1] == crc8(inner, 0, inner.count - 1) else { return nil }
        return Data(inner[magic.count ..< inner.count - 1])
    }

    private static func crc8(_ data: [UInt8], _ from: Int, _ to: Int) -> UInt8 {
        var crc = 0
        for i in from ..< to {
            crc ^= Int(data[i])
            for _ in 0 ..< 8 {
                crc = (crc & 0x80) != 0 ? ((crc << 1) ^ 0x07) & 0xFF : (crc << 1) & 0xFF
            }
        }
        return UInt8(crc)
    }

    private static func tokenize(_ text: String) -> [Substring] { StegoTokenizer.runs(text) }
}
